### PR TITLE
Adding decorators support

### LINKED ISSUE #143: Conversion fails when syntax uses decorators
```
error processing blabla/blabla.js
{ SyntaxError: This experimental syntax requires enabling one of the following parser plugin(s): 'decorators-legacy, decorators' (7:2)
    at Object.raise (/home/eek/.config/yarn/global/node_modules/@babel/parser/lib/index.js:7012:17)
    at Object.expectOnePlugin (/home/eek/.config/yarn/global/node_modules/@babel/parser/lib/index.js:8420:18)
    at Object.parseDecorator (/home/eek/.config/yarn/global/node_modules/@babel/parser/lib/index.js:10872:10)
    at withTopicForbiddingContext (/home/eek/.config/yarn/global/node_modules/@babel/parser/lib/index.js:11475:32)
    at Object.withTopicForbiddingContext (/home/eek/.config/yarn/global/node_modules/@babel/parser/lib/index.js:10565:14)
    at Object.parseClassBody (/home/eek/.config/yarn/global/node_modules/@babel/parser/lib/index.js:11464:10)
    at Object.parseClass (/home/eek/.config/yarn/global/node_modules/@babel/parser/lib/index.js:11438:22)
    at Object.parseStatementContent (/home/eek/.config/yarn/global/node_modules/@babel/parser/lib/index.js:10732:21)
    at Object.parseStatement (/home/eek/.config/yarn/global/node_modules/@babel/parser/lib/index.js:10690:17)
    at Object.parseStatement (/home/eek/.config/yarn/global/node_modules/@babel/parser/lib/index.js:2079:26)
  pos: 124,
  loc: Position { line: 7, column: 2 },
  missingPlugin: [ 'decorators-legacy', 'decorators' ] }
```

Flow to TS seems to fail if the script uses decorators. 

Is there a way I can enable this babel plugin?

### PR DESCRIPTION
**Pull request by aminya**
This adds decorators support such as for mobx. I have added the test which was taken from a [real-world project](https://github.com/nteract/hydrogen/blob/cdd0ba3526cdf82eddba8e538133a8c7c747687d/lib/store/index.js#L29).

Fixes #143

### CODE REVIEW

**Comment by aminya** [comment]
@kevinbarabash Would you please check this out? 😃

File: .prettierignore [hunk 1]
```diff
@@ -2,6 +2,7 @@ test/fixtures/convert/formatting
 test/fixtures/convert/imports/named_import_specifier_kind/
 test/fixtures/convert/function-types/predicate03
 test/fixtures/convert/imports/dynamic_import
+test/fixtures/convert/decorator/
```
Comment (aminya) [inline .prettierignore:5]: Just for the record, this was added because prettier removed the `@observable` from above the `markersMapping` in the test when the parser [is flow](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgAmHwgCMBnOHANwEMSAbOAGn0iwFd4ATFmsGVNHwBffADMcEdPgA6IdKQAecgNwyocRdjyt6NMmXwBlGLjiF1+fAAFSFanUaX86GjgDWlMgFkamTKhQAOZI+L6YADxQ7OgklCy+HpQmZgB8+AC8+BoA7mF+ABQAlGpQ6sLqIEwgEJgC0GTIoG6SOQAKbgiNKDRUEKhcVSAkOHyeMEaYfIFByDA47MwgcLFwXFxrADI0wew0QXAAYriuMALByCA0nBBDABYw6PQA6neo8GRTYHBGXe+oVHeAE9LmADENAvYYG1RkFXMgxDR6BRqgArMiKABCozA4yMNHQcE2gTgCKRKJA6MURhmjAAiuwIPAycillMcPZLmJ6BAckNMDhAjBngMYHdkAAOAAM1QFEAoz1GmEuArg9iopOqAEdGfAYbVulcyABaDRrNZDHBwHWoK0w-bwpCI1nVCjoVBzBZLMi0uAMpmkp3kpYwOgirhi5AAJmq8xoqHoMwAwlJHSAoNBNSB2BQACp0brOilURYASSgG1gRjAgrqAEEKyYgYwWRT9DBW0suBAwJ3qhBOJhOFHO8JhEA).

Fortunately, it [works fine with TypeScript](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEBLAtgBwgJxgAmHwgCMBnOHANwEMSAbOAGn0iwFd4ATFmsGVNHwBffADMcEdPgA6IdKQAecgNwyocRdjyt6NMmXwBlGLjiF1+fAAFSFanUaX86GjgDWlMgFkamTKhQAOZI+L6YADxQ7OgklCy+HpQmZgB8+AC8+BoA7mF+ABQAlGpQ6sLqIEwgEJgC0GTIoG6SOQAKbgiNKDRUEKhcVSAkOHyeMEaYfIFByDA47MwgcLFwXFxrADI0wew0QXAAYriuMALByCA0nBBDABYw6PQA6neo8GRTYHBGXe+oVHeAE9LmADENAvYYG1RkFXMgxDR6BRqgArMiKABCozA4yMNHQcE2gTgCKRKJA6MURhmjAAiuwIPAycillMcPZLjAgZg4GQwDhUHUhphBbBngMYHdkAAOAAM1VFEAoz1GmEuor5lCopOqAEdGfAYbVulcyABaDRrNZDHBwA2oO0w-bwpCI1nVCjoVBzBZLMi0uAMpmkt3kpYwOgSrhS5AAJmq8xoqHoMwAwlJXSAoNBdSB2BQACp0bruilURYASSgG1gRgFQpgAEEayYgYwWRT9DBO0suBAwL3qhBOJhOHHe8JhEA), which is what flow-to-ts [uses in the end](https://github.com/Khan/flow-to-ts/blob/10beb7d9dc256d533efd642f2b118df83d7267b9/src/convert.ts#L86) for formatting with prettier.

File: package.json [hunk 2]
```diff
@@ -49,6 +51,7 @@
     "webpack-cli": "^4.5.0"
   },
   "scripts": {
+    "prepare": "git submodule update --init --recursive",
```
Comment (k4b7) [inline package.json]: Thanks for adding this.  Hopefully checking out `babel` as a submodule isn't too much of a pain.

File: package.json [hunk 3]
```diff
@@ -36,6 +37,7 @@
     "@types/typescript": "^2.0.0",
     "babel-loader": "^8.2.2",
     "babel-plugin-module-resolver": "^4.1.0",
+    "charcodes": "^0.2.0",
```
Comment (k4b7) [inline package.json]: Why was this added?

File: test/fixtures/convert/decorator/mobx/flow.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    return savedFilePath ? savedFilePath : `Unsaved Editor ${editor.id}`;
+  }
+
+  @action
+  newMarkerStore(editorId: number) {
+    const markerStore = new MarkerStore();
+    this.markersMapping.set(editorId, markerStore);
+    return markerStore;
+  }
+}
```
Comment (k4b7) [inline test/fixtures/convert/decorator/mobx/flow.js]: I like that this test case covers a few different uses of decorators, but the implementation details add a bunch of noise that is distracting when comparing the input to the output.  Could you pare this test case down to the bare essentials?

**Review by k4b7 — COMMENTED** [review]
Thanks for the PR.  It's cool that transform.ts didn't need to be changed.  There are maybe a few more test cases that might be worth adding:
- multiple decorators
- decorator on a class
- decorator factories

**Comment by dimitropoulos** [comment]
Just want to chime in that I would benefit quite a lot from this in the conversion of insomnia: https://github.com/Kong/insomnia/issues/3266

For now, we'll be using a minimal implementation of this PR https://github.com/dimitropoulos/flow-to-ts/commit/41a52dc915d23d1b01955de046adc963d52128ee (I made it a few days ago before this PR existed).

We'll need to be using this code in our conversion so it'd be useful if it could be merged soon.

I'd be happy to help however possible.

File: package.json (hunk 2 above)
Comment (aminya) [reply package.json]: At first, I didn't notice that the repo was using submodules, and I was getting some errors about the missing folder, and I finally found out that the repo is using submodules. 😄 Adding this script fixes this for future contributors.

File: package.json (hunk 3 above)
Comment (aminya) [reply package.json]: That is a dependency used in one of the Babel packages. Because we added the Babel folder without bootstrapping it, we need to keep track of their dependencies here.

The alternative is to bootstrap those packages.

**Comment by aminya** [comment]
> Just want to chime in that I would benefit quite a lot from this in the conversion of insomnia: [Kong/insomnia#3266](https://github.com/Kong/insomnia/issues/3266)
> 

@dimitropoulos Could you give us some simple examples of the decorators you use and the expected output so we can add them to the tests?

**Comment by aminya** [comment]
> Thanks for the PR. It's cool that transform.ts didn't need to be changed. There are maybe a few more test cases that might be worth adding:
> 
> * multiple decorators
> * decorator on a class
> * decorator factories

Done!

**Comment by dimitropoulos** [comment]
Screenshot_20210407_164205

Thankfully we only have exactly one decorator, `autoBindMethodsForReact` from https://www.npmjs.com/package/class-autobind-decorator, that we use in 148 places.

To pick a random example: https://github.com/Kong/insomnia/blob/develop/packages/insomnia-app/app/ui/components/editors/password-editor.js#L24

**Comment by aminya** [comment]
> Thankfully we only have exactly one decorator, `autoBindMethodsForReact` from [npmjs.com/package/class-autobind-decorator](https://www.npmjs.com/package/class-autobind-decorator), that we use in 148 places.
> 

OK. I added the test for that one too.

File: package.json (hunk 3 above)
Comment (k4b7) [reply package.json]: Running `yarn install` in the `babel` directory should be enough.  That should probably be added to the `prepare` script you added instead adding individual babel dependencies here.

File: test/fixtures/convert/decorator/mobx/flow.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+  @computed
+  get filePath(): ?string {
+    return "/file";
+  }
+
+  @action
+  newMarkerStore(editorId: number) {
+    return 1;
+  }
+}
```
Comment (k4b7) [inline test/fixtures/convert/decorator/mobx/flow.js:15]: Thanks for simplifying these examples.  The could be simpler since the code doesn't have to be semantically valid (in this case the getter doesn't return a value, e.g.
```
class MyClass {
   @property
   num: number = 5;

   @computed
   get foo(): string {}

   @action
   doSomething() {}
}
```
What you have is fine for now.  No action req'd.

File: test/fixtures/convert/decorator/decorator-on-class-with-argument/flow.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import { autoBindMethodsForReact } from "class-autobind-decorator";
+
+@autoBindMethodsForReact(AUTOBIND_CFG)
+class PasswordEditor extends React.PureComponent<Props, State> {
```
Comment (k4b7) [inline test/fixtures/convert/decorator/decorator-on-class-with-argument/flow.js:4]: It seems like the `decorator-factory` example already covers this, unless I'm missing something.

Comment (aminya) [reply test/fixtures/convert/decorator/decorator-on-class-with-argument/flow.js:4]: I can remove it if you prefer fewer tests, but usually, more tests are better 😄

File: package.json (hunk 3 above)
Comment (aminya) [reply package.json]: Done
https://github.com/Khan/flow-to-ts/pull/271/commits/23ced32d7359837994fef21610518dcf6d4722b1

File: test/fixtures/convert/decorator/decorator-on-class-with-argument/flow.js (hunk 6 above)
Comment (k4b7) [reply test/fixtures/convert/decorator/decorator-on-class-with-argument/flow.js:4]: It's fine.  It'd be a different story if the tests took a long time to run, but that isn't the case.

**Review by k4b7 — APPROVED** [review]
LGTM.  Thanks for adding this feature. 🙂

**Comment by k4b7** [comment]
I've published 0.5.1 to npm.

**Comment by dimitropoulos** [comment]
wow this is great news, now I don't have to publish a fork just for our use-case!  thank you all so much!
